### DISCUSSION TITLE
Using a current sensor to calibrate a roller shutter

### DISCUSSION THREAD

**Discussion by shmuelzon** [arendst/Tasmota#12143] [Ideas]
Hello,

From what I understood from the [documentation](https://tasmota.github.io/docs/Blinds-and-Shutters/), in order to calibrate shutters and blinds, one needs to measure the time it takes to open and close the shutter and configure that in.
I was wondering if that could be automated using a current sensor, like in the [Sonoff Dual R3](https://sonoff.tech/product/diy-smart-switch/dualr3/), assuming the motor has limit switches.

The calibration process would roll the shutter up until the motor doesn't pull any current and define that as fully open, then roll the shutter all the way down and measure how long it takes until the motor hits the lower limit switch, i.e. stops pulling current again. We can roll it up once more to verify the results are similar, or just average the two.

One would still need to set the half-way point, if needed, but I believe that this procedure would greatly simplify the calibration process.

What do you think?

Thanks!

**Comment by sfromis** [comment]
I'm afraid that precise timing would not work out due to the slower rate of energy monitoring. I'd prefer having the limit switches wired up to the Tasmota device, but of course this may not be practical.

**Reply by shmuelzon** [reply]
In my specific case, the limits are inside the motor itself and are set via some sort of potentiometer. I'm guessing it just counts its own turns and no actual switches are involved.

Regardless, what's the current rate of energy monitoring? Is it a software limit in Tasmota itself or a limit in the CSE7761 found on the Sonoff Dual R3?
I'm also wondering what kind of accuracy is actually needed here. Surely even 100ms intervals would be better than what most would measure manually with a stopwatch and `ShutterEnableEndStopTime<X>` would help limit the drift over time.

**Reply by sfromis** [reply]
Technically, those energy monitoring chips are not current monitors. They measure a more complex combo, including hertz and waveforms, to also produce numbers for apparent/reactive power and power factor. I've not pored over maximum capabilities of specific chips, but from cheap IoT devices I'd not expect higher sampling rate than seconds. Energy monitoring is a very different purpose than a current measurement instrument. 

Anyway, the Tasmota main loop for polling sensors is something like a 1Hz rate.

**Reply by sfromis** [reply]
When it comes to manual measurements, absolute accuracy is not important there, as people can easily say "oh, I need a bit longer time" to get the the wanted position.

**Reply by shmuelzon** [reply]
While I wasn't able to find an actual demo of it, it's my understanding that the stock Sonoff firmware, at least for the R3, has this capability and can be activated from the mobile app with a "calibrate motor" button, see [here](https://www.youtube.com/watch?v=BDc9uROtJus&t=828s), so the hardware should be capable. It doesn't need to be very accurate but just detect a sharp drop in consumption.

In terms of implementation, I'm guessing the firmware would need to be put in some special mode where it would poll the sensor more often or even run as a separate task when calibrating but I'm not familiar with code at all so I don't know what that would actually require, if at all possible with the current architecture.

Anyways, I'm putting it out there as a suggestion / feature request.
Should I have the time to implement it myself, are there any pointers you can suggest?

**Comment by TrDA-hab** [comment]
1. In the very concept of Tasmota and its current implementation, your proposal (idea / proposal) is absolutely not acceptable.
2. I'm sorry, but your task requires specialized and highly targeted firmware.
3. I spent a lot of time implementing "Shutter Mode" from Tasmota (2019-2021) and spent a lot of my and other people's time.
4. Now it works great and does not require any additions.
**Good luck!**

**Reply by shmuelzon** [reply]
Hey @TrDA-hab,

I feel like I've offended you and I deeply apologize if that's the case.
I have nothing bad to say about Tasmota in general nor the current shutter implementation in particular and was only looking for a way to automate a manual task and, theoretically, reach a more accurate calibration value.

Like I said before, I'm not familiar with the Tasmota architecture but I assumed that it can check if a current sensor/measurement is available and make use of it so I did not think anything specialized is actually needed, "just" and additional action that can be invoked by a user which will make use of entities (relays + sensors) already supported.

Anyways, if this feature is as unacceptable as you say it is, feel free to mark this discussion as such.

Thanks
